toolchain/check/context.cpp

```diff
@@ -227,22 +228,35 @@ auto Context::ResolveIfImportRefUnused(SemIR::InstId inst_id) -> void {
   }
   const SemIR::File& import_ir = *import_irs().Get(unused_inst->ir_id);
   auto import_inst = import_ir.insts().Get(unused_inst->inst_id);
-  switch (import_inst.kind()) {
-    default:
-      // TODO: We need more type support. For now we inject an arbitrary
-      // invalid node that's unrelated to the underlying value. The TODO
-      // diagnostic is used since this section shouldn't typically be able to
-      // error.
-      TODO(Parse::NodeId::Invalid,
-           (llvm::Twine("TODO: ResolveIfImportRefUnused for ") +
-            import_inst.kind().name())
-               .str());
-      ReplaceInstBeforeConstantUse(
-          inst_id,
-          {SemIR::ImportRefUsed{SemIR::TypeId::Error, unused_inst->ir_id,
-                                unused_inst->inst_id}});
-      break;
+
+  // If the type ID isn't a normal value, forward it directly.
+  if (!import_inst.type_id().is_valid()) {
+    ReplaceInstBeforeConstantUse(
+        inst_id,
+        {SemIR::ImportRefUsed{import_inst.type_id(), unused_inst->ir_id,
+                              unused_inst->inst_id}});
+    return;
   }
+
+  auto import_type_inst_id = import_ir.types().GetInstId(import_inst.type_id());
+  CARBON_CHECK(import_type_inst_id.is_valid());
+
+  // If the type of the instruction is a builtin, use it directly.
+  auto type_id = SemIR::TypeId::Invalid;
```

Comment:
I'd avoid the dead initialization. Sanitizers and static analysis tools seem useful to catch bugs here.

```suggestion
  SemIR::TypeId type_id;
```

Reply:
So far we haven't been supporting default construction of ID types, so the suggestion doesn't compile. We've been relying more on preventing accidental default construction, since use-cases seem relatively rare (we have about a dozen `Invalid` inits right now). Maybe #toolchain is a good place to discuss if you'd like to add default construction?